rfcs/20220610-saved-model-fingerprinting.md

```diff
@@ -0,0 +1,293 @@
+   </td>
+   <td><strong>Deterministic?</strong>
+   </td>
+   <td><strong>Content-based?</strong>
+   </td>
+   <td><strong>Unique?</strong>
+   </td>
+   <td><strong>Useful for:</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>File Checksum
+   </td>
+   <td>Enable tracking of the specific savedmodel
+   </td>
+   <td>No
+   </td>
+   <td>Yes
+   </td>
+   <td>Yes
+   </td>
+   <td>Tracking specific protobufs in production, such as those available on TF Hub
+   </td>
+  </tr>
+  <tr>
+   <td>Regularized hashes
+   </td>
+   <td>Enable tracking of the source code
+   </td>
+   <td>Yes
+   </td>
+   <td>Yes
+   </td>
+   <td>Yes
+   </td>
+   <td>Library authors who want to know who is using their code
+   </td>
+  </tr>
+</table>
+
+
+## Comparison of the File Checksum and Program Hashes
+
+What parts of the fingerprint change?
+
+
+<table>
+  <tr>
+   <td><strong>Case</strong>
+   </td>
+   <td><strong>File Checksum</strong>
+   </td>
+   <td><strong>Program Hashes</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>Model A and Model B are produced by calling `tf.saved_model.save(model, “/tmp)` function twice in a row in the same process.
+   </td>
+   <td>Different
+   </td>
+   <td>Same
+   </td>
+  </tr>
+  <tr>
+   <td>Model A and Model B are produced by the exact same program using `bazel run`.
+   </td>
+   <td>Different
+   </td>
+   <td>Same
+   </td>
+  </tr>
+  <tr>
+   <td>Model B is a copy of Model A (e.g. you ctrl-C, ctrl-V the file).
+   </td>
+   <td>Same
+   </td>
+   <td>Same
+   </td>
+  </tr>
+</table>
+
+
+## Design Proposal
+
+The engineering impact of this project is significant in that it adds an additional file (`fingerprint.pb`) to the SavedModel format.
+
+
+### A Separate File for the Fingerprint
+
+We plan to add an additional protobuf in the SavedModel directory:
+
+*   `assets/`
+*   `saved_model.pb`
+*   **<code>fingerprint.pb</code>**
+*   <code>variables/</code>
+
+The advantages of having a separate file for the fingerprint are:
+- It neatly encapsulates the changes that this design proposes
+- It does not interfere with the MetaGraphDef in the saved_model.pb. The MetaGraphDef protobuf is already extremely large so we want to avoid adding fields to it.
+
+One drawback of having a separate file is that it is a potentially significant change to the SavedModel format and could break any users who are using the SavedModel format in ways not endorsed by our library.
+
+### API for reading and writing the fingerprints
+
+All fingerprints will be written to disk automatically, and there is no way to bypass or customize writing fingerprints when serializing a SavedModel.
+
+To read the fingerprint, users can load the protobuf into memory and read the field. At first, there will not be dedicated APIs in Python to get the fingerprint, since no other part of the SavedModel has special APIs to access its value. However, if the need arises, we can add APIs under the `tf.saved_model` namespace to read the fingerprint.
```

Comment:
Could it be in protobuf text format so that we could have the fingerprint  human readable?

Reply:
That's a good idea. However, neither the `saved_model.pb` nor the `saved_model_debug_info.pb` are protobuf texts, so for consistency I would keep the `fingerprint.pb` in serialized form.

Reply:
Ok, added a section under "Alternatives Considered".